src/server/session.ts

```diff
@@ -2198,7 +2204,23 @@ namespace ts.server {
             const scriptInfo = project.getScriptInfoForNormalizedPath(file)!;
             const { startPosition, endPosition } = this.getStartAndEndPosition(args, scriptInfo);
 
-            const codeActions = project.getLanguageService().getCodeFixesAtPosition(file, startPosition, endPosition, args.errorCodes, this.getFormatOptions(file), this.getPreferences(file));
+            let codeActions: readonly CodeFixAction[];
+            try {
+                codeActions = project.getLanguageService().getCodeFixesAtPosition(file, startPosition, endPosition, args.errorCodes, this.getFormatOptions(file), this.getPreferences(file));
+            }
+            catch(e) {
+                const existingDiagCodes = flatten((["getSyntacticDiagnostics", "getSemanticDiagnostics", "getSuggestionDiagnostics"] as const).map(getter =>
+                    this.getDiagnosticsWorker(args, /*isSemantic*/ getter !== "getSyntacticDiagnostics",
+                                              (project, file) => project.getLanguageService()[getter](file),
+                                              /*includeLinePosition*/ true)
+                        .filter(d => decodedTextSpanIntersectsWith(startPosition, endPosition - startPosition, d.start, d.length))
+                        .map(d => d.code)));

```

Comment:
It doesn’t seem like `this.getDiagnosticsWorker` is buying you anything (which makes its signature change unnecessary). You don’t use the line position or the formatting work done there. You can just use the diagnostics you get directly from the language service. Also, the inner `filter` and `map` can be replaced with a single `mapDefined` and the outer `flatten` and `map` can be replaced with a single `flatMap`. Alternatively, it may be clearer just to do

```ts
const existingDiagCodes = mapDefined([
  ...languageService.getSyntacticDiagnostics(file),
  ...languageService.getSemanticDiagnostics(file),
  ...languageService.getSuggestionDiagnostics(file)
], d => {
  // ... if intersecting, return code
});

Reply:
* Re `flatMap` — I missed it, thanks!

* Re using `getDiagnosticsWorker` — it was certainly painful to go through it, but it has a suspicious use of `isDeclarationFileInJSOnlyNonConfiguredProject` which I thought would be good to preserve (not only for whatever it's doing now, but in case more such tweaks are added later on).  If it's fine to use the diagnostics directly, then it's definitely simplifying things.

* Re `definedMap` — I missed that one also, but actually there's no need for that since it's fine to just use an unfiltered `map`...

So the current version uses the diagnostics directly.

Reply:
####- [ ] - _****_@_[]()_